src/lib.rs

```diff
@@ -78,3 +78,75 @@ pub fn from_value<T: serde::de::DeserializeOwned>(value: JsValue) -> Result<T> {
 pub fn to_value<T: serde::ser::Serialize + ?Sized>(value: &T) -> Result<JsValue> {
     value.serialize(&Serializer::new())
 }
+
+/// Serialization and deserialization functions that pass JavaScript objects through unchanged.
+///
+/// This module is compatible with the `serde(with)` annotation, so for example if you create
+/// the struct
+///
+/// ```rust
+/// #[derive(serde::Serialize)]
+/// struct MyStruct {
+///     int_field: i32,
+///     #[serde(with = "serde_wasm_bindgen::preserve")]
+///     js_field: js_sys::Int8Array,
+/// }
+///
+/// let s = MyStruct {
+///     int_field: 5,
+///     js_field: js_sys::Int8Array::new_with_length(1000),
+/// };
+/// ```
+///
+/// then `serde_wasm_bindgen::to_value(&s)`
+/// will return a JsValue representing an object with two fields (`int_field` and `js_field`), where
+/// `js_field` will be an `Int8Array` pointing to the same underlying JavaScript object as `s.js_field` does.
+pub mod preserve {
+    use serde::{de::Error, Deserialize, Serialize};
+    use wasm_bindgen::{
+        convert::{FromWasmAbi, IntoWasmAbi},
+        JsCast, JsValue,
+    };
+
+    // Some arbitrary string that no one will collide with unless they try.
+    pub(crate) const PRESERVED_VALUE_MAGIC: &str = "1fc430ca-5b7f-4295-92de-33cf2b145d38";
+
+    #[derive(Serialize, Deserialize)]
+    #[serde(rename = "1fc430ca-5b7f-4295-92de-33cf2b145d38")]
+    struct PreservedValueWrapper(u32);
+
+    /// Serialize any `JsCast` value.
+    ///
+    /// When used with the `Serializer` in `serde_wasm_bindgen`, this serializes the value by
+    /// passing it through as a `JsValue`.
+    ///
+    /// This function is compatible with the `serde(serialize_with)` derive annotation.
+    pub fn serialize<S: serde::Serializer, T: JsCast>(val: &T, ser: S) -> Result<S::Ok, S::Error> {
+        // Since we don't own `val` we need to clone it. Otherwise serializing a JsValue
+        // will produce a second JsValue pointing to the same index in the wasm-bindgen heap,
+        // and whichever one is dropped first will leave the other one dangling.
+        PreservedValueWrapper(val.as_ref().clone().into_abi()).serialize(ser)
+    }
+
+    /// Deserialize any `JsCast` value.
+    ///
+    /// When used with the `Derializer` in `serde_wasm_bindgen`, this serializes the value by
+    /// passing it through as a `JsValue` and casting it.
+    ///
+    /// This function is compatible with the `serde(deserialize_with)` derive annotation.
+    pub fn deserialize<'de, D: serde::Deserializer<'de>, T: JsCast>(de: D) -> Result<T, D::Error> {
+        let wrap = PreservedValueWrapper::deserialize(de)?;
```

Comment:
I think what still concerns me particularly is that if you look at what `derive(Deserialize)` does on a newtype struct, it simply generates a call like `deserializer.deserialize_newtype_struct("1fc430ca-5b7f-4295-92de-33cf2b145d38", visitor)` and expects the deserializer to invoke `visit_newtype_struct` with the nested value.

However, most deserializers completely ignore the newtype and just fall-through to the nested value. Case in point, serde-json:

https://github.com/serde-rs/json/blob/8cebe895007bd1305228f21d5fe6d200e136c4f3/src/de.rs#L1712-L1713

That is, this will happily accept any `u32` from any deserializer without actually validating the name. As you said, this might be not a UB, but it still seems like a relatively easy way to shoot oneself in a foot and get some arbitrary JS value without clarity on what went wrong.

FWIW something similar happens in serialize stage - e.g. https://github.com/serde-rs/json/blob/8cebe895007bd1305228f21d5fe6d200e136c4f3/src/ser.rs#L228 - but, of course, that one is less dangerous direction.

I think it's better to change the repr to some more robust option that could fail reliably even with non-self-descriptive format like JSON. Some options:

```rust
enum PreservedValue {
  #[serde(rename="(guid)")]
  Abi(u32),
}
```

or

```rust
struct PreservedValue {
  #[serde(rename="(guid)")]
  abi: u32,
}
```

In both cases the representation in JSON would look like `{"(guid)": 42}`, that is, the GUID will be embedded and strictly checked upon deserialization.

Reply:
Oh, that's a good point. Actually it occurred to me that we may also want to do something to prevent round-tripping in other serializer/deserializers. For example, even with the repr you proposed I think serializing with `serde_json`, sending to another process, and deserializing with `serde_json` will result in a "successful" deserialization that makes up a bogus `JsValue`.

One way to prevent this would be to have some mismatch between the ser/de parts. For example, we could have

```rust
#[serde(rename="(guid)")]
struct PreservedValueSer(u32);

struct PreservedValueDe {
  #[serde(rename="(guid)")]
  abi: u32,
}
```

Our serialize function in the `with` module converts a `JsValue` to a `PreservedValueSer` and serializes it, while the deserialize function deserializes a `PreservedValueDe` and converts to a `JsValue`. Most serializer/deserializers will fail to round-trip this, but our serializer would recognize the guid of the newtype struct and serialize it as a `PreservedValueDe` instead.

(I think this still might round-trip with something like bincode which (IIRC) ignores struct field names, so we could even do something like add an extra field to `PreservedValueDe`)

Reply:
Sounds like a good idea, sure. The `PreservedValueSer` struct to be the same as `PreservedValueDe` - that is, non-newtype, otherwise we're back to the same problem.

> (I think this still might round-trip with something like bincode which (IIRC) ignores struct field names, so we could even do something like add an extra field to PreservedValueDe)

Yeah bincode is harder... I'm inclined to say let's ignore it for now, I think it's far less commonly used than serde-json (via the old wasm-bindgen's serde integration) on Wasm-related structs. But, if you have a good solution that doesn't affect performance, go for it.